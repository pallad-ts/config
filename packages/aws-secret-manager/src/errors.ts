import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

export const errorsDomain = new Domain();

const code = formatCodeFactory("E_CONF_SECRET_MANAGER_%c");
export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
	CANNOT_EXTRACT_PROPERTY_FROM_NON_OBJECT: ErrorDescriptor.useDefaultMessage(
		code(1),
		"Cannot extract property from non-object"
	),
	MISCONFIGURED_DATALOADER: ErrorDescriptor.useMessageFormatter(
		code(2),
		(batchSize: number) => `Misconfigured dataloader. Missing required "maxBatchSize: ${batchSize}" option`
	),
});
