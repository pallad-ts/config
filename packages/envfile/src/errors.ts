import {Domain, generators} from 'alpha-errors';
import {ConfigError} from "@pallad/config";

export const ERRORS = new Domain({
    errorClass: ConfigError,
    codeGenerator: generators.formatCode('E_CONF_ENVFILE_%d')
}).createErrors(create => {
    return {
        ENV_FILE_DOES_NOT_EXIST: create('Env file %s does not exist but it is required'),
    };
});
