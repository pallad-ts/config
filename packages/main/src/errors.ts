import {Domain, generators} from 'alpha-errors';
import {ConfigError} from "./ConfigError";

export const ERRORS = new Domain({
    errorClass: ConfigError,
    codeGenerator: generators.formatCode('E_CONF_%d')
}).createErrors(create => {
    return {
        CONFIG_LOADING_FAILED: create('Config loading failed: \s%s'),
        PROVIDER_VALUE_NOT_AVAILABLE: create('Value not available: %s'),
        SYNC_LOADING_NOT_AVAILABLE: create('Cannot load config synchronously since one of your config value providers is asynchronous'),
        PICK_PROVIDER_UNREGISTERED_TYPE: create('Unregistered type "%s". Registered values: %s'),
        PICK_PROVIDER_TYPE_ALREADY_EXISTS: create('Type "%s" already registered'),
        CANNOT_CONVERT_TO_INT: create(`Value "%s" cannot be converted to int`),
        CANNOT_CONVERT_TO_NUMBER: create(`Value "%s" cannot be converted to number`),
        CANNOT_CONVERT_TO_BOOL: create(`Value "%s" cannot be converted to bool`),
        TYPE_URL_INVALID_PROTOCOL: create('Protocol "%s" is not allowed. Allowed: %s'),
        INVALID_PORT_NUMBER: create('Port "%s" is invalid. Allowed ranges: %s')
    };
});
