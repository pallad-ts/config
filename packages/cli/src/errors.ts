import { ConfigError } from "@pallad/config";
import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

const errorsDomain = new Domain();
const code = formatCodeFactory("E_CONF_CLI_%c");

export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
    INVALID_CONFIG_VALUE_AT_PATH: ErrorDescriptor.useMessageFormatter(
        code(1),
        (parentConfigValue: unknown, propertyPath: string[]) =>
            `Cannot get config at property "${propertyPath.join(",")}" from config since parent value is of type: ${typeof parentConfigValue}`,
        ConfigError
    ),
});
