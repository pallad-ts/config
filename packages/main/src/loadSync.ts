import { ERRORS } from "./errors";
import { loadConfig } from "./common/loadConfig";
import { isPromise } from "./common/isPromise";
import { handleConfigLoadResult } from "./common/handleConfigLoadResult";
import { ResolvedConfig } from "./ResolvedConfig";

/**
 * Loads config synchronously
 *
 * If config contains any asynchronous provider then loading fails
 *
 * @public
 */
export function loadSync<T extends object>(config: T): ResolvedConfig<T> {
    const result = loadConfig(config);
    if (isPromise(result)) {
        throw ERRORS.SYNC_LOADING_NOT_AVAILABLE.create();
    }
    return handleConfigLoadResult(result);
}
