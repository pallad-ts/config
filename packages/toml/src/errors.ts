import { ConfigError } from "@pallad/config";
import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

const errorsDomain = new Domain();
const code = formatCodeFactory("E_CONF_TOML_%c");

export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
    ENV_FILE_DOES_NOT_EXIST: ErrorDescriptor.useMessageFormatter(
        code(1),
        (filePath: string) => `TOML file "${filePath}" does not exist but it is required`,
        ConfigError
    ),
});
