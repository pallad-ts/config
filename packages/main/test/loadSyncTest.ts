import {loadSync} from '@src/loadSync';
import {DummyProvider} from './dummies/DummyProvider';
import {ERRORS} from '@src/errors';
import {Either} from 'monet';

describe('loadSync', () => {
    it('loads config', () => {
        const input = {
            foo: DummyProvider.sync({value: 'foo'}),
            nested: {
                bar: DummyProvider.sync({value: 'bar'})
            }
        };

        const config = loadSync(input);
        expect(config)
            .toStrictEqual({
                foo: 'foo',
                nested: {
                    bar: 'bar'
                }
            });
    });

    describe('fails', () => {
        it('if contains async provider', () => {
            expect(() => {
                loadSync({
                    foo: DummyProvider.async({value: 'foo'})
                })
            })
                .toThrowErrorMatchingSnapshot();
        });

        it('if some of providers is not available', () => {
            const result = Either.fromTry(() => {
                loadSync({
                    foo: DummyProvider.notAvailable({isAsync: false, description: 'foo desc'})
                })
            });
            expect(result.left())
                .toHaveProperty('code', ERRORS.CONFIG_LOADING_FAILED.code);

            expect(result.left())
                .toHaveProperty('errors', [
                    ERRORS.PROVIDER_VALUE_NOT_AVAILABLE.format('foo desc')
                ]);
        });

        it('if some of providers fails', () => {
            const error = new Error('foo');
            const result = Either.fromTry(() => {
                loadSync({
                    foo: DummyProvider.notAvailable({isAsync: false, error})
                })
            });
            expect(result.left())
                .toHaveProperty('code', ERRORS.CONFIG_LOADING_FAILED.code);

            expect(result.left())
                .toHaveProperty('errors', [
                    error
                ]);
        });

        it('if some of providers fails with multiple errors', () => {
            const error1 = new Error('test');
            const error2 = new Error('e2');
            const error3 = new Error('e3');
            const error4 = new Error('e4');

            const result = Either.fromTry(() => {
                loadSync({
                    foo: new DummyProvider({isAsync: false, error: [error1, error2]}),
                    var: new DummyProvider({isAsync: false, error: [error3, error4]})
                })
            });

            expect(result.left())
                .toHaveProperty('code', ERRORS.CONFIG_LOADING_FAILED.code);

            expect(result.left())
                .toHaveProperty('errors', [
                    error1,
                    error2,
                    error3,
                    error4
                ]);
        });
    });
});
