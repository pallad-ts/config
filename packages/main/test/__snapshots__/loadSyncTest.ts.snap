// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadSync fails if contains async provider 1`] = `"Cannot loadConfig config synchronously since one of your config value provider is asynchronous"`;

